lib/phoenix_live_view/html_formatter.ex

```diff
@@ -0,0 +1,451 @@
+  #
+  # Will be tokenized as:
+  #
+  # [
+  #   {:tag_open, "section", [], %{column: 1, line: 1}},
+  #   {:text, "\n  ", %{column_end: 3, line_end: 2}},
+  #   {:tag_open, "p", [], %{column: 3, line: 2}},
+  #   {:eex_tag_render, "<%= user.name ></p>\n  <%= if true do %>", %{block?: true, column: 6, line: 1}},
+  #   {:text, " ", %{column_end: 2, line_end: 1}},
+  #   {:tag_open, "p", [], %{column: 2, line: 1}},
+  #   {:text, "deu bom", %{column_end: 12, line_end: 1}},
+  #   {:tag_close, "p", %{column: 12, line: 1}},
+  #   {:eex_tag, "<% else %>", %{block?: false, column: 35, line: 2}},
+  #   {:tag_open, "p", [], %{column: 1, line: 1}},
+  #   {:text, " deu ruim ", %{column_end: 14, line_end: 1}},
+  #   {:tag_close, "p", %{column: 14, line: 1}},
+  #   {:eex_tag, "<% end %>", %{block?: false, column: 62, line: 2}},
+  #   {:text, "\n", %{column_end: 1, line_end: 2}},
+  #   {:tag_close, "section", %{column: 1, line: 2}}
+  # ]
+  #
+  # EEx.tokenize/2 was introduced in Elixir 1.14.
+  # TODO: Remove this when we no longer support earlier versions.
+  @eex_expr [:start_expr, :expr, :end_expr, :middle_expr]
+  if Code.ensure_loaded?(EEx) && function_exported?(EEx, :tokenize, 2) do
+    defp tokenize(contents) do
+      {:ok, eex_nodes} = EEx.tokenize(contents)
+      {tokens, cont} = Enum.reduce(eex_nodes, {[], :text}, &do_tokenize/2)
+      HTMLTokenizer.finalize(tokens, "nofile", cont)
+    end
+
+    defp do_tokenize({:text, text, _meta}, {tokens, cont}) do
+      text
+      |> List.to_string()
+      |> HTMLTokenizer.tokenize("nofile", 0, [], tokens, cont)
+    end
+
+    defp do_tokenize({:comment, text, meta}, {tokens, cont}) do
+      {[{:eex_comment, List.to_string(text), meta} | tokens], cont}
+    end
+
+    defp do_tokenize({type, opt, expr, %{column: column, line: line}}, {tokens, cont})
+         when type in @eex_expr do
+      meta = %{opt: opt, line: line, column: column}
+      {[{:eex, type, expr |> List.to_string() |> String.trim(), meta} | tokens], cont}
+    end
+  else
+    defp tokenize(contents) do
+      {:ok, eex_nodes} = EEx.Tokenizer.tokenize(contents, 1, 0, %{indentation: 0, trim: false})
+      {tokens, cont} = Enum.reduce(eex_nodes, {[], :text}, &do_tokenize/2)
+      HTMLTokenizer.finalize(tokens, "nofile", cont)
+    end
+
+    defp do_tokenize({:text, _line, _column, text}, {tokens, cont}) do
+      text
+      |> List.to_string()
+      |> HTMLTokenizer.tokenize("nofile", 0, [], tokens, cont)
+    end
+
+    defp do_tokenize({type, line, column, opt, expr}, {tokens, cont}) when type in @eex_expr do
+      meta = %{opt: opt, line: line, column: column}
+      {[{:eex, type, expr |> List.to_string() |> String.trim(), meta} | tokens], cont}
+    end
+  end
+
+  defp do_tokenize(_node, acc) do
+    acc
+  end
+
+  # Build an HTML Tree according to the tokens from the EEx and HTML tokenizers.
```

Comment:
@josevalim not sure if we want to keep these comments. There were useful when I was developing but maybe we can drop it

Reply:
I would keep it!